File: tests/test_ranges.py
import pytest
from gto.cards import Card
from gto.ranges import (
    combo_count, parse_range, range_from_top_pct, total_combos,
    range_pct, blockers_remove, blocked_combos,
)


class TestComboCount:
    def test_pair(self):
        assert combo_count("AA") == 6

    def test_suited(self):
        assert combo_count("AKs") == 4

    def test_offsuit(self):
        assert combo_count("AKo") == 12


class TestParseRange:
    def test_simple(self):
        result = parse_range("AA,KK,QQ")
        assert "AA" in result
        assert "KK" in result
        assert "QQ" in result

    def test_plus_pairs(self):
        result = parse_range("TT+")
        assert "TT" in result
        assert "JJ" in result
        assert "QQ" in result
        assert "KK" in result
        assert "AA" in result
        assert "99" not in result

    def test_plus_suited(self):
        result = parse_range("ATs+")
        assert "ATs" in result
        assert "AJs" in result
        assert "AQs" in result
        assert "AKs" in result
        assert "A9s" not in result

    def test_dash_pairs(self):
        result = parse_range("77-TT")
        assert "77" in result
        assert "88" in result
        assert "99" in result
        assert "TT" in result
        assert "66" not in result
        assert "JJ" not in result

    def test_dash_suited(self):
        result = parse_range("KTs-KQs")
        assert "KTs" in result
        assert "KJs" in result
        assert "KQs" in result
        assert "K9s" not in result

    def test_mixed(self):
        result = parse_range("AA, KK, AKs, AQs+")
        assert "AA" in result
        assert "AKs" in result


class TestRangeFromTopPct:
    def test_top_1(self):
        result = range_from_top_pct(1)
        assert "AA" in result
        assert len(result) <= 3

    def test_top_10(self):
        result = range_from_top_pct(10)
        combos = total_combos(result)
        assert combos <= 1326 * 0.12

    def test_top_50(self):
        result = range_from_top_pct(50)
        assert len(result) > 20

    def test_invalid(self):
        with pytest.raises(ValueError):
            range_from_top_pct(0)
        with pytest.raises(ValueError):
            range_from_top_pct(101)


class TestTotalCombos:
    def test_basic(self):
        assert total_combos(["AA"]) == 6
        assert total_combos(["AA", "KK"]) == 12
        assert total_combos(["AKs"]) == 4
        assert total_combos(["AKo"]) == 12


class TestRangePct:
    def test_all_pairs(self):
        pairs = ["AA", "KK", "QQ", "JJ", "TT", "99", "88", "77", "66", "55", "44", "33", "22"]
        pct = range_pct(pairs)
        assert abs(pct - (78 / 1326 * 100)) < 0.1


class TestBlockers:
    def test_remove_blocked(self):
        hero = [Card("A", "s"), Card("K", "h")]
        result = blockers_remove(["AA", "KK", "QQ"], hero)
        assert "AA" in result  # still has some combos
        assert "QQ" in result  # not blocked at all

    def test_blocked_count(self):
        hero = [Card("A", "s"), Card("A", "h")]
        count = blocked_combos("AA", hero)
        assert count == 5  # holding 2 aces blocks 5 of 6 combos

File: tests/test_ranges.rs
use gto_cli::cards::*;
use gto_cli::ranges::*;

#[test]
fn test_combo_count_pair() {
    assert_eq!(combo_count("AA"), 6);
}

#[test]
fn test_combo_count_suited() {
    assert_eq!(combo_count("AKs"), 4);
}

#[test]
fn test_combo_count_offsuit() {
    assert_eq!(combo_count("AKo"), 12);
}

#[test]
fn test_parse_range_simple() {
    let result = parse_range("AA,KK,QQ");
    assert!(result.contains(&"AA".to_string()));
    assert!(result.contains(&"KK".to_string()));
    assert!(result.contains(&"QQ".to_string()));
}

#[test]
fn test_parse_range_plus_pairs() {
    let result = parse_range("TT+");
    assert!(result.contains(&"TT".to_string()));
    assert!(result.contains(&"JJ".to_string()));
    assert!(result.contains(&"QQ".to_string()));
    assert!(result.contains(&"KK".to_string()));
    assert!(result.contains(&"AA".to_string()));
    assert!(!result.contains(&"99".to_string()));
}

#[test]
fn test_parse_range_plus_suited() {
    let result = parse_range("ATs+");
    assert!(result.contains(&"ATs".to_string()));
    assert!(result.contains(&"AJs".to_string()));
    assert!(result.contains(&"AQs".to_string()));
    assert!(result.contains(&"AKs".to_string()));
    assert!(!result.contains(&"A9s".to_string()));
}

#[test]
fn test_parse_range_dash_pairs() {
    let result = parse_range("77-TT");
    assert!(result.contains(&"77".to_string()));
    assert!(result.contains(&"88".to_string()));
    assert!(result.contains(&"99".to_string()));
    assert!(result.contains(&"TT".to_string()));
    assert!(!result.contains(&"66".to_string()));
    assert!(!result.contains(&"JJ".to_string()));
}

#[test]
fn test_parse_range_dash_suited() {
    let result = parse_range("KTs-KQs");
    assert!(result.contains(&"KTs".to_string()));
    assert!(result.contains(&"KJs".to_string()));
    assert!(result.contains(&"KQs".to_string()));
    assert!(!result.contains(&"K9s".to_string()));
}

#[test]
fn test_parse_range_mixed() {
    let result = parse_range("AA, KK, AKs, AQs+");
    assert!(result.contains(&"AA".to_string()));
    assert!(result.contains(&"AKs".to_string()));
}

#[test]
fn test_range_from_top_pct_1() {
    let result = range_from_top_pct(1.0).unwrap();
    assert!(result.contains(&"AA".to_string()));
    assert!(result.len() <= 3);
}

#[test]
fn test_range_from_top_pct_10() {
    let result = range_from_top_pct(10.0).unwrap();
    let combos = total_combos(&result);
    assert!(combos <= (1326.0 * 0.12) as u32);
}

#[test]
fn test_range_from_top_pct_50() {
    let result = range_from_top_pct(50.0).unwrap();
    assert!(result.len() > 20);
}

#[test]
fn test_range_from_top_pct_invalid() {
    assert!(range_from_top_pct(0.0).is_err());
    assert!(range_from_top_pct(101.0).is_err());
}

#[test]
fn test_total_combos() {
    assert_eq!(total_combos(&["AA".to_string()]), 6);
    assert_eq!(
        total_combos(&["AA".to_string(), "KK".to_string()]),
        12
    );
    assert_eq!(total_combos(&["AKs".to_string()]), 4);
    assert_eq!(total_combos(&["AKo".to_string()]), 12);
}

#[test]
fn test_range_pct_all_pairs() {
    let pairs: Vec<String> = vec![
        "AA", "KK", "QQ", "JJ", "TT", "99", "88", "77", "66", "55", "44", "33", "22",
    ]
    .into_iter()
    .map(String::from)
    .collect();
    let pct = range_pct(&pairs);
    assert!((pct - (78.0 / 1326.0 * 100.0)).abs() < 0.1);
}

#[test]
fn test_blockers_remove() {
    let hero = vec![
        Card::new(Rank::Ace, Suit::Spades),
        Card::new(Rank::King, Suit::Hearts),
    ];
    let range: Vec<String> = vec!["AA", "KK", "QQ"]
        .into_iter()
        .map(String::from)
        .collect();
    let result = blockers_remove(&range, &hero);
    assert!(result.contains(&"AA".to_string())); // still has some combos
    assert!(result.contains(&"QQ".to_string())); // not blocked at all
}

#[test]
fn test_blocked_combos_count() {
    let hero = vec![
        Card::new(Rank::Ace, Suit::Spades),
        Card::new(Rank::Ace, Suit::Hearts),
    ];
    let count = blocked_combos("AA", &hero).unwrap();
    assert_eq!(count, 5); // holding 2 aces blocks 5 of 6 combos
}
